### PR TITLE
Design API v3

### LINKED ISSUE #10: TASK: Design API
To be able to use this XbSymbolDatabase in Cxbx-Reloaded (and possible other projects, like via an IDA Pro plugin), we'll (obviously) be needing an API to access it's functionality.

Things that this library should be able to do :
* query the library for it's version & capabilities
* register symbol-table database(s) [ this might not be needed until #8 is decided upon ]
* start a scanning session, returning a handle
* scan a memory-range for symbols (internally adding the detected symbols to the session handle)
* retrieve the detected symbols in a session
* closing a scanning sessions, freeing all allocated resources

This API must be implemented with thread-safety in mind.
API signatures must be designed such, that future updates to the library could be added without breaking the interface.

The API that will be doing the actual scanning, must be able to be updated in the future, so it might be wise to give it only one argument : a structure pointer that passes in all arguments.
This way, the structure can be changed in the future, without breaking the BAPI.

Things that should go into the scanning struct:
* a struct version
* scanning flags (or a reserved DWORD to put those in once we need them)
* an indication of the range / ranges that should be scanned
  * per range: it's size in bytes, a pointer to it's data, and it's base-address in virtual-memory

Please discuss further below

### PR DESCRIPTION
**Pull request by RadWolfie**
Design API v3.0

Closes #10

New features: (see XbSymbolDatabase.h file for functions' details)
* XbSymbolDatabase_GenerateLibraryFilter
* XbSymbolDatabase_GenerateSectionFilter
* XbSymbolDatabase_CreateXbSymbolContext
* XbSymbolContext_ScanLibrary (design for multi-thread safe support)
   * If there is a dependency on another library's symbols. Such as XACTENG library do depend on DSOUND's symbols. Then second or later libraries will not perform the scan until the library is done scanning in the active thread.
* XbSymbolContext_ScanAllLibraryFilter (design for single-threading only)
   * May will change to support multi-thread usage internally.
* XbSymbolContext_RegisterXRefs (currently is a stub function)
  * Can be resolved by #39
* XbSymbolContext_Release

Currently, the design is backward compatibility. Until final stage will move the old APIs into new APIs.

### CODE REVIEW

**Comment by RadWolfie** [comment]
Currently, multi-thread support is not enabled for MSVC due to the developers haven't implemented C11 support in their compiler. From what I had read in the recent past, they're working on it and should be bundle in future Visual Studio 2019 update.

**Comment by RadWolfie** [comment]
Plus most importantly, include pointer to kernel thunk's address to translate internally for make stronger OOVPA signatures easier. 😄

**Comment by RadWolfie** [comment]
If the team want to make the final decision for API's names later, please let me know here or on discord. Otherwise, we need to continue discussion what's need to be add/remove/update the parameters and/or API's names.

File: XbSymbolDatabase.h [hunk 1]
```diff
@@ -447,6 +834,33 @@ static const char* xbe_type_str[XB_XBE_TYPE_MAX] = {
...
+    Library*        filters;
+} LibraryHeader;
+
+typedef struct _Section {
+    // These three variables must be fixed info.
+    char            name[8];
+    uint32_t        xb_virt_addr;
+    uint32_t        buffer_size;
+    void*           buffer_lower;
+} Section;
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: These very generic names should get prefixes to avoid conflicts with other APIs / user-code.

Ideally this was true (and consistent) for *all* types / variables / functions in the public headers.

File: XbSymbolDatabase.h [hunk 2]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
 /// <param name="enable">Input boolean to use first symbol address only or not.</param>
 void XbSymbolFirstDetectAddressOnly(bool enable);
 
+// ******************************************************************
+// * API v3 (context) functions to use with other projects.
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: Why even keep the old API? People who don't want to use the new API can just check out the old version (at loss of features), or make independent wrappers.
Porting apps is probably easy eitherway (and there aren't many users to begin with).

File: XbSymbolDatabase.h [hunk 3]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
+
+/// <summary>
+/// Step 3: Create XbSymbolContext context for the scan process.
+/// </summary>
+/// <param name="ppContext">Output XbSymbolContext handler.</param>
+/// <param name="register_func">Register callback function for any detected symbols during scan process.</param>
+/// <param name="library_input">See XbSymbolGenerate_LibraryFilter for details.</param>
+/// <param name="section_input">See XbSymbolGenerate_SectionFilter for details.</param>
+/// <param name="kernel_thunk">Input decoded virtual kernel thunk.</param>
+bool XbSymbol_Init(XbSymbolContext* ppContext, xb_symbol_register_t register_func, LibraryHeader library_input, SectionHeader section_input, xbaddr kernel_thunk);
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: Why do I `XbSymbol_Init` to *create*?
And why does `Init` happen as step 3?
This implies the name is bad.

File: XbSymbolDatabase.h [hunk 4]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
+
+/// <summary>
+/// Step 4: Perform a manual scan to set xreferences manually by requirement.
+/// </summary>
+/// <param name="pContext">Input XbSymbolContext handler.</param>
+void XbSymbolScan_Manual(XbSymbolContext pContext);
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: These new prefixes are weird?
Functions with different prefix interact; but unlike COM, there don't seem to be objects involved.

I'd expect this to be `XbSymbolContext_ScanManual` (`Object_Method`) or `XbSymbolScanManual` (`PrefixMethod` - existing style) or `XbSymbol_ScanManual` ('Module_Method')

File: XbSymbolDatabase.h [hunk 5]
```diff
@@ -447,6 +834,33 @@ static const char* xbe_type_str[XB_XBE_TYPE_MAX] = {
...
+    uint32_t        buffer_size;
+    void*           buffer_lower;
+} Section;
+
+typedef struct _SectionHeader {
+    unsigned int    count;
+    Section*        filters;
+} SectionHeader;
+
+typedef void* XbSymbolContext;
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: This should be renamed to suggest that it is a handle or reference.

Current name implies a struct / class / object; so I'd expect code to work like this (which is incorrect):

```c
XbSymbolContext* ctx = malloc(sizeof(XbSymbolContext));
XbSymbol_Init(&XbSymbolContext, ...);
```

or (also incorrect):

```c
XbSymbolContext* ctx = XbSymbol_CreateContext(...);
```

In C, you usually assume a type to be an object that you don't want to assign / pass by value (as it might be too big). So if you have a non-struct-like-type, something should imply that it's a number / pointer of some sort.

File: XbSymbolDatabase.h [hunk 6]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
 void XbSymbolFirstDetectAddressOnly(bool enable);
 
+// ******************************************************************
+// * API v3 (context) functions to use with other projects.
+// ******************************************************************
+// NOTE: If need to make customize setup, please use the above functions
+// before call XbSymbol_Init function.
+
+/// <summary>
+/// Step 1: Generate library array for LibraryHeader input.
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: Ideally the API would imply the order of calls, so we don't have to document it.

File: XbSymbolDatabase.h (hunk 4 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: True, I will fix it later on in this pull request.

File: XbSymbolDatabase.h (hunk 3 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: There are necessary steps must be done first before pass on two data arrays. But yeah, I see your point for whole name is bad. It is still possible to manually create both data arrays on your own if using static library method.

File: XbSymbolDatabase.h (hunk 6 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: Well, if use a document tool such as doxygen will generate a documentation for human readability as a help/research tool. Beside, if hover above the API name also provide details too.

Comment (RadWolfie) [reply XbSymbolDatabase.h]: Well, if use a document tool such as doxygen will generate a documentation for human readability as a help/research tool. Beside, if hover above the API name also provide details too.

File: XbSymbolDatabase.h (hunk 4 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: fixed

File: XbSymbolDatabase.h (hunk 5 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: fixed

File: XbSymbolDatabase.h [hunk 7]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
+bool XbSymbolContext_Create(XbSymbolContextHandle* ppHandle, xb_symbol_register_t register_func, XbSDBLibraryHeader library_input, XbSDBSectionHeader section_input, xbaddr kernel_thunk);
+
+/// <summary>
+/// Step 4: Perform a manual scan to set xreferences manually by requirement.
+/// </summary>
+/// <param name="pHandle">Input XbSymbolContextHandle handler.</param>
+void XbSymbolContext_ScanManual(XbSymbolContextHandle pHandle);
+
+/// <summary>
+/// Step 5: (multi-thread) Process individual library input by third-party.
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: What makes this multi-thread?
Is it just thread-safe, or does it actually spawn threads internally?

And why is this part of the "Step" instead of generic information in the summary?

File: XbSymbolDatabase.h [hunk 8]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
+/// </summary>
+/// <param name="pHandle">Input XbSymbolContextHandle handler.</param>
+/// <param name="pLibrary">Input pointer of a library to start a scan process.</param>
+/// <param name="xref_first_pass">True: Optimized first search scan process; False: 2nd call and later will continue to return non-zero; see return for detail.</param>
+/// <returns>Return total xref count found. If return zero, then there's nothing left to find.</returns>
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: > If return zero, then there's nothing left to find.

This isn't grammatically correct.
Should this be or something like this:

> If it returns zero, then there's nothing left to find.

Could be made even better with more changes.

File: XbSymbolDatabase.h [hunk 9]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
+unsigned int XbSymbolContext_ScanLibrary(XbSymbolContextHandle pHandle, const XbSDBLibrary* pLibrary, bool xref_first_pass);
+
+/// <summary>
+/// Step 5: (single-thread) Process all of the libraries internally.
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: The single-thread note feels unnecessary.

File: XbSymbolDatabase.h [hunk 10]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
+/// </summary>
+/// <param name="pHandle">Input XbSymbolContextHandle handler.</param>
+void XbSymbolContext_ScanFull(XbSymbolContextHandle pHandle);
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: Maybe rename to `ScanAllLibraries` if that's what it does (or does it do anything else, other than scanning libs?).

File: XbSymbolDatabase.h [hunk 11]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
+
+/// <summary>
+/// Step 6: Register any xreferences may not had been output during the scan process.
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: Note: Sometimes you use xreferences (here) and sometimes xrefs.
For comments, I'd suggest calling them "cross-references" for clarity.

File: XbSymbolDatabase.h (hunk 6 above)
Comment (JayFoxRox) [reply XbSymbolDatabase.h]: I still think this implies bad design.

People can also learn the order by looking at the code sample, or it will be implied by what each function is doing (which implies the order).
For example, you can't possibly do: `XbSymbolContext_ScanManual` if you didn't create an object of type `XbSymbolContext` before (using `XbSymbolContext_Create` - which is a misnomer at the moment). This implies an order without having to explicitly document it.

In doxygen, you wouldn't be putting the step information near each function. Instead, you'd keep the generic documentation (order of calls) separated.

File: XbSymbolDatabase.h [hunk 12]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
+
+/// <summary>
+/// Step 3: Create XbSymbolContextHandle context for the scan process.
+/// </summary>
+/// <param name="ppHandle">Output XbSymbolContextHandle handler.</param>
+/// <param name="register_func">Register callback function for any detected symbols during scan process.</param>
+/// <param name="library_input">See XbSymbolDatabase_GenerateLibraryFilter for details.</param>
+/// <param name="section_input">See XbSymbolDatabase_GenerateSectionFilter for details.</param>
+/// <param name="kernel_thunk">Input decoded virtual kernel thunk.</param>
+bool XbSymbolContext_Create(XbSymbolContextHandle* ppHandle, xb_symbol_register_t register_func, XbSDBLibraryHeader library_input, XbSDBSectionHeader section_input, xbaddr kernel_thunk);
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: If you go with the `Object_Action` naming scheme, then this is a misnomer.
You'd want to call this `XbSymbolDatabase_CreateXbSymbolContext` or something (so it creates the object which can be used later to operate on).

File: XbSymbolDatabase.h [hunk 13]
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
...
+unsigned int XbSymbolDatabase_GenerateLibraryFilter(const void* xb_header_addr, XbSDBLibraryHeader* library_out);
+
+/// <summary>
+/// Step 2: Generate section array for SectionHeader input.
+/// First call with <paramref name="section_out"/> as null pointer will return total count. Then second call will insert information to <paramref name="section_out"/>.filters field.
+/// </summary>
+/// <param name="xb_header_addr"></param>
+/// <param name="section_out">Input null pointer will provide the count to allocate memory, then allocate memory to section_out.filters field.</param>
+/// <param name="is_raw">True: Convert array to use raw xbe's relative address; False: Convert array to use xbox virtual memory relative address.</param>
+unsigned int XbSymbolDatabase_GenerateSectionFilter(const void* xb_header_addr, XbSDBSectionHeader* section_out, bool is_raw);
```
Comment (JayFoxRox) [inline XbSymbolDatabase.h]: This stuff (filters) could probably be treated similar to the function which creates a context (returning a handle and share the naming conventions: `Create` or something).

Basically this would imply that you have to create the filters first, in order to be able to create the context (as it depends on the filters). Then you can operate on the context.

So the API will eventually become self-documenting.

File: XbSymbolDatabase.h (hunk 12 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: Sorry, some of my changes got lost in the process. It's suppose to be `XbSymbolContextCreate`.

File: XbSymbolDatabase.h (hunk 7 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: Multi-thread safe, if the compiler support C11 standard. Currently, msvc compiler doesn't have C11 standard support.

File: XbSymbolDatabase.h (hunk 10 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: It's more of a scan all filter libraries. The proper wording for me would be `ScanAllFilterLibs`.

File: XbSymbolDatabase.h (hunk 11 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: "x" wasn't used for "cross", it was actually "xbox" internal usage in the past codebase. And it still sort of is, I'll extend the comment about this.

File: XbSymbolDatabase.h (hunk 6 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: Right now, there's no sample in this repo. Nor has a tool to for scan xbe files. That's the main reason steps are in the documentations. Later down the road, all of the steps comment will be remove.

File: XbSymbolDatabase.h (hunk 12 above)
Comment (JayFoxRox) [reply XbSymbolDatabase.h]: Even that name is somewhat unstandard and fragments the API further.

The proposed name would make this the only function that is *not* of style `Object_Action`.
It would also lead to the old issue with overlapping prefixes (`XbSymbol` and `XbSymbolContext`).

If we keep the old API, too (we shouldn't.. but this PR doesn't seem to remove it yet), then this is even more confusing, as that already used such naming conventions.

I'm not sure how a user could figure out `XbSymbolContextCreate` without looking at the headers (and even then it's confusing to me).

As a user, I'd have expected the call I have mentioned (if the rest of the API uses objects or a singleton).

File: XbSymbolDatabase.h (hunk 13 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: I'm avoiding handle for both header's structures. Since I don't have a valid reason for one time usage which is only used in create process. And it's easier to reduce extra workload for simple allocation and set the new temporary buffer once. Once `XbSymbolContextCreate` have a copy of the filter's structure data. Then it is safe to free temporary buffer. Unless use the alternate function for multi-threading support, then you'll have to provide library info to `XbSymbolContext_ScanLibrary`.

Plus, I don't plan on implement a reference counter system.

File: XbSymbolDatabase.h (hunk 12 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: 👌

File: XbSymbolDatabase.h (hunk 9 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: It's use for awareness, beside I have reword it to understand better.

File: XbSymbolDatabase.h (hunk 2 above)
Comment (RadWolfie) [reply XbSymbolDatabase.h]: Updated in the comment section.

**Comment by RadWolfie** [comment]
Since the API naming and comments was changed since initialize commit. I do know initialize commit is working and had been verified using my unit test project locally.

**Comment by RadWolfie** [comment]
What's been change:

Added:
* XbSymbolDatabase_GetKernelThunkAddress

Renamed:
* XbSymbolContext_RegisterLibrary
* XbSymbolContext_SetBypassBuildVersionLimit
* XbSymbolContext_SetContinuousSigScan
* XbSymbolContext_SetFirstDetectAddressOnly
* XbSymbolDatabase_LibraryVersion
* XbSymbolDatabase_SetOutputMessage
* XbSymbolDatabase_LibraryToFlag
* XbSymbolDatabase_TestOOVPAs
* XbSymbolDatabase_SetOutputVerbosity

---

Internal functions has been renamed in second commit.

Plus added:
* iXbSymbolContext_Lock
* iXbSymbolContext_Unlock
* iXbSymbolContext_AllowSetParameter
* iXbSymbolContext_AllowScanLibrary

**Comment by RadWolfie** [comment]
**NOTE:** `XbSymbolScan` function has not been remove yet. It is now deprecated function.

File: XbSymbolDatabase.c [hunk 14]
```diff
@@ -84,59 +89,97 @@ static inline uint32_t BitScanReverse(uint32_t value)
...
+// doing the same scan process.
+typedef enum _eLibraryType {
+    LT_D3D = 0,
+    LT_AUDIO,
+    LT_JVS,
+    LT_XAPI,
+    LT_GRAPHIC,
+    LT_NETWORK,
+    LT_COUNT,
+    LT_UNKNOWN=LT_COUNT
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: It's cleaner to swap these two, so list LT_UNKNOWN first, and declare LT_COUNT as an assignment of LT_UNKNOWN.

File: XbSymbolDatabase.c [hunk 15]
```diff
@@ -359,21 +488,190 @@ uint32_t XbSymbolLibraryToFlag(const char* library_name)
...
+                        output_message_format(&output, XB_OUTPUT_MESSAGE_WARN,
+                            "D3D8 version 1.0.%d.%d Title Detected: This game uses an alias version 1.0.5788",
+                            xb_library_versions[library_index].wBuildVersion,
+                            xb_library_versions[library_index].wFlags.QFEVersion);
+
+                        library_header->filters[count].build_version = 5788;
+                    }
+                }
+            }
+            count++;
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: This causes the `filters` array to have gaps. If that's undesirable, increase `count` only when populating a filter.

Same happens below, I'll mark them with a single `gap` comment.

File: XbSymbolDatabase.c [hunk 16]
```diff
@@ -359,21 +488,190 @@ uint32_t XbSymbolLibraryToFlag(const char* library_name)
...
+        // Manual check if DSOUND section do exist.
+        if (!has_dsound_library) {
+            if (XbSymbolHasDSoundSection(xb_header_addr)) {
+                if (library_header != NULL) {
+                    library_header->filters[count].flag = XbSymbolLib_DSOUND;
+                    library_header->filters[count].build_version = xb_library_versions[library_index].wBuildVersion;
+                    library_header->filters[count].qfe_version = 0;
+                    memcpy(library_header->filters[count].name, Lib_DSOUND, 8);
+                }
+                count++;
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: `gap`

File: XbSymbolDatabase.c [hunk 17]
```diff
@@ -359,21 +488,190 @@ uint32_t XbSymbolLibraryToFlag(const char* library_name)
...
+
+        // Manual check if Xbe type is debug or Chihiro
+        if (XbeType != XB_XBE_TYPE_RETAIL) {
+            if (library_header != NULL) {
+                library_header->filters[count].flag = XbSymbolLib_JVS;
+                library_header->filters[count].build_version = xb_library_versions[library_index].wBuildVersion;
+                library_header->filters[count].qfe_version = 0;
+                memcpy(library_header->filters[count].name, Lib_JVS, 8);
+            }
+            count++;
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: `gap`

File: XbSymbolDatabase.c [hunk 18]
```diff
@@ -359,21 +488,190 @@ uint32_t XbSymbolLibraryToFlag(const char* library_name)
...
+
+                        if (is_raw) {
+                            sv_index->buffer_lower = (memptr_t)xb_header_addr + sh_index->dwRawAddr;
+                        }
+                        else {
+                            sv_index->buffer_lower = xb_start_addr + sh_index->dwVirtualAddr;
+                        }
+                    }
+
+                    count++;
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: `gap`

File: XbSymbolDatabase.c [hunk 19]
```diff
@@ -84,59 +89,97 @@ static inline uint32_t BitScanReverse(uint32_t value)
...
-
-    OOVPATable     *OovpaTable;
+    OOVPATable*     OovpaTable;
     unsigned int    OovpaTableCount;
 } SymbolDatabaseList;
 
 SymbolDatabaseList SymbolDBList[] = {
     // Support inline functions in .text section
-    { XbSymbolLib_D3D8 | XbSymbolLib_D3D8LTCG,{ Sec_text, Sec_D3D, Sec_FLASHROM }, D3D8_OOVPA, D3D8_OOVPA_COUNT },
+    { LT_D3D, XbSymbolLib_D3D8 | XbSymbolLib_D3D8LTCG,{ Sec_text, Sec_D3D, Sec_FLASHROM }, D3D8_OOVPA, D3D8_OOVPA_COUNT },
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: Why include a library type enum value here, when there's also a `internal_GetLibraryType` function that's doing the same? Either remove this variable, and use the function, or remove the function (or implement it as a lookup in SymbolDBList, instead of switch/case code)

File: XbSymbolDatabase.c [hunk 20]
```diff
@@ -658,39 +956,113 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
...
+            return LT_XAPI;
+        case XbSymbolLib_XGRAPHC:
+            return LT_GRAPHIC;
+        case XbSymbolLib_XONLINE:
+        case XbSymbolLib_XONLINES:
+        case XbSymbolLib_XNET:
+        case XbSymbolLib_XNETS:
+        case XbSymbolLib_XNETN:
+            return LT_NETWORK;
+    }
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: Remove this function, or implement it as a lookup in `SymbolDBList` (instead of switch/case code).

File: XbSymbolDatabase.c [hunk 21]
```diff
@@ -762,158 +1134,253 @@ bool XbSymbolScanSection(uint32_t xbe_base_address,
...
+    pContext->section_input.count = section_input.count;
+    pContext->section_input.filters = malloc(sizeof(XbSDBSection) * section_input.count);
+    if (pContext->section_input.filters == NULL) {
+        goto LibraryCleanup;
+    }
+
+    memcpy(pContext->library_input.filters, library_input.filters, sizeof(XbSDBLibrary) * library_input.count);
+    memcpy(pContext->section_input.filters, section_input.filters, sizeof(XbSDBSection) * section_input.count);
+
+                            // Mark all Xrefs initially as undetermined
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: nit : indentation

File: XbSymbolDatabase.h [hunk 22]
```diff
@@ -91,12 +91,393 @@ extern "C" {
...
+    KT_IMP_FUNC_XProfpControl,                          // 0x0172 (370) PROFILING
+    KT_IMP_FUNC_XProfpGetData,                          // 0x0173 (371) PROFILING
+    KT_IMP_FUNC_IrtClientInitFast,                      // 0x0174 (372) PROFILING
+    KT_IMP_FUNC_IrtSweep,                               // 0x0175 (373) PROFILING
+    KT_IMP_FUNC_MmDbgAllocateMemory,                    // 0x0176 (374) DEVKIT ONLY!
+    KT_IMP_FUNC_MmDbgFreeMemory,                        // 0x0177 (375) DEVKIT ONLY!
+    KT_IMP_FUNC_MmDbgQueryAvailablePages,               // 0x0178 (376) DEVKIT ONLY!
+    KT_IMP_FUNC_MmDbgReleaseAddress,                    // 0x0179 (377) DEVKIT ONLY!
+    KT_IMP_FUNC_MmDbgWriteCheck,                        // 0x017A (378) DEVKIT ONLY!
+    KT_IMP_COUNT,
```
Comment (PatrickvL) [inline XbSymbolDatabase.h]: Instead of `KT_IMP_` prefix, use `XREF_KT_` prefix.

File: XbSymbolDatabase.c [hunk 23]
```diff
@@ -1430,178 +1869,252 @@ bool XbLibraryScan(custom_scan_func_t custom_scan_func,
...
+    XbSymbolContextHandle pHandle;
+    iXbSymbolContext* iContext;
+    XbSDBLibraryHeader library_input;
+    XbSDBSectionHeader section_input;
+
+    // Step 1, let's get the total sum of array to allocate library input.
+    library_input.count = XbSymbolDatabase_GenerateLibraryFilter(xb_header_addr, NULL);
+    // If total sum is zero, then the input is invalid.
+    if (library_input.count == 0) {
+        goto EmptyCleanup;
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: Avoid goto, just return here.

File: XbSymbolDatabase.c [hunk 24]
```diff
@@ -1430,178 +1869,252 @@ bool XbLibraryScan(custom_scan_func_t custom_scan_func,
...
+    }
+    // Then allocate the necessary memory requirement to obtain the information.
+    library_input.filters = malloc(sizeof(XbSDBLibrary) * library_input.count);
+    // Verify if memory has been allocated.
+    if (library_input.filters == NULL) {
+        goto EmptyCleanup;
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: Avoid goto, just return here.

File: XbSymbolDatabase.c [hunk 25]
```diff
@@ -658,39 +956,113 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
...
+        case XbSymbolLib_XNETN:
+            return LT_NETWORK;
+    }
+}
+
+xbaddr XbSymbolDatabase_GetKernelThunkAddress(const void* xb_header_addr)
+{
+    xb_xbe_type xbe_type = GetXbeType(xb_header_addr);
+
+    uint32_t kernel_thunk_addr = *(uint32_t*)((uint8_t*)xb_header_addr + 0x0158);
```
Comment (PatrickvL) [inline XbSymbolDatabase.c]: Use a symbol instead of the 'magic' literal hexadecimal number `0x0158`.

File: XbSymbolDatabase.c (hunk 19 above)
Comment (RadWolfie) [reply XbSymbolDatabase.c]: Okay, I'll need to take a step back to see what you're seeing. I'm sure your suggestion method is better. 😁

File: XbSymbolDatabase.c (hunk 15 above)
Comment (RadWolfie) [reply XbSymbolDatabase.c]: I think I see what you mean, I'll re-review this function again.

File: XbSymbolDatabase.c (hunk 25 above)
Comment (RadWolfie) [reply XbSymbolDatabase.c]: I didn't make any modification here except move into its own function.

File: XbSymbolDatabase.c (hunk 19 above)
Comment (RadWolfie) [reply XbSymbolDatabase.c]: After reviewing again, obviously this variable is no longer needed. 😃

File: XbSymbolDatabase.c (hunk 20 above)
Comment (RadWolfie) [reply XbSymbolDatabase.c]: Since SymbolDBList could be used two or more times in a single library scan. This function, `internal_GetLibraryType` will stay to keep the scan process active once. 😉

File: XbSymbolDatabase.c (hunk 15 above)
Comment (RadWolfie) [reply XbSymbolDatabase.c]: I don't see the problem. I think you had missed the API's documentation. If `library_header` is null pointer, it will return the counts for what is needed to allocate filter's array. Then re-call the function again to fill in the filters.

**Comment by RadWolfie** [comment]
Fixed all reviews by @PatrickvL except one which is working as intended.

Plus I discover a warning message from Visual Studio about memcpy usage over the limit. I have replaced memcpy to strncpy function for prevent warning and null'd the overflow.
